vowpalwabbit/namespaced_feature_store.h

```diff
@@ -4,253 +4,217 @@
+  using const_iterator = indexed_iterator_t<const std::array<std::list<namespaced_features>, 256>,
+      std::vector<namespace_index>::const_iterator, const std::list<namespaced_features>>;
+  using list_iterator = std::list<namespaced_features>::iterator;
+  using const_list_iterator = std::list<namespaced_features>::const_iterator;
+  using index_flat_iterator = VW::chained_proxy_iterator<list_iterator, features::audit_iterator>;
+
+  namespaced_feature_store() = default;
+  ~namespaced_feature_store() = default;
+
+  // For copies we deep copy the features but not the pool
+  namespaced_feature_store(const namespaced_feature_store& other)
+  {
+    _feature_groups = other._feature_groups;
+    _legacy_indices_existing = other._legacy_indices_existing;
+  }
+
+  namespaced_feature_store& operator=(const namespaced_feature_store& other)
+  {
+    _feature_groups = other._feature_groups;
+    _legacy_indices_existing = other._legacy_indices_existing;
+    return *this;
+  }
+  namespaced_feature_store(namespaced_feature_store&& other) = default;
+  namespaced_feature_store& operator=(namespaced_feature_store&& other) = default;
+
+  inline size_t size() const
+  {
+    size_t accumulator = 0;
+    for (auto ns_index : _legacy_indices_existing) { accumulator += _feature_groups[ns_index].size(); }
+    return accumulator;
+  }
+  inline bool empty() const { return _legacy_indices_existing.empty(); }
 
   // Returns nullptr if not found.
-  features* get_feature_group(uint64_t hash);
-  // Returns nullptr if not found.
-  const features* get_feature_group(uint64_t hash) const;
+  inline features* get_or_null(namespace_index ns_index, uint64_t hash)
+  {
+    auto& bucket = _feature_groups[ns_index];
+    auto it = std::find_if(
+        bucket.begin(), bucket.end(), [hash](const namespaced_features& group) { return group.hash == hash; });
+    if (it == bucket.end()) { return nullptr; }
+    return &it->feats;
+  }
 
-  const std::set<namespace_index>& get_indices() const;
-  namespace_index get_index_for_hash(uint64_t hash) const;
+  // Returns nullptr if not found.
+  inline const features* get_or_null(namespace_index ns_index, uint64_t hash) const
+  {
+    auto& bucket = _feature_groups[ns_index];
+    auto it = std::find_if(
+        bucket.begin(), bucket.end(), [hash](const namespaced_features& group) { return group.hash == hash; });
+    if (it == bucket.end()) { return nullptr; }
+    return &it->feats;
+  }
 
-  // The following are experimental and may be superseded with namespace_index_begin_proxy
-  // Returns empty range if not found
-  std::pair<indexed_iterator, indexed_iterator> get_namespace_index_groups(namespace_index ns_index);
-  // Returns empty range if not found
-  std::pair<const_indexed_iterator, const_indexed_iterator> get_namespace_index_groups(namespace_index ns_index) const;
+  inline features& get(namespace_index ns_index, uint64_t hash)
+  {
```

Comment:
so we should only call `get` if we are 100% sure that the ns/hash has features, otherwise we can still call `get_or_null`

Reply:
Correct